sdk/nodejs/output.ts

```diff
@@ -312,6 +236,77 @@ To manipulate the value of this Output, use '.apply' instead.`);
             },
         });
     }
+
+    public get(): T {
+        throw new Error(`Cannot call '.get' during update or preview.
+To manipulate the value of this Output, use '.apply' instead.`);
+    }
+
+    public apply<U>(func: (t: T) => Input<U>): Output<U> {
+        const applied: Promise<OutputData<U>> =
+            Promise.all([this.promise(), this.isKnown, this.isSecret])
+                   .then(([value, isKnown, isSecret]) => applyHelperAsync<T, U>(value, isKnown, isSecret, func));
+
+        const result = new OutputImpl<U>(
+            this.resources(),
+            applied.then(a => a.value),
+            applied.then(a => a.isKnown),
+            applied.then(a => a.isSecret));
+        return <Output<U>><any>result;
+    }
+}
+
+interface OutputData<T> {
+    value: T;
+    isKnown: boolean;
+    isSecret: boolean;
+}
+
+async function applyHelperAsync<T, U>(value: T, isKnown: boolean, isSecret: boolean, func: (t: T) => Input<U>): Promise<OutputData<U>> {
+    if (runtime.isDryRun()) {
+        // During previews only perform the apply if the engine was able to
+        // give us an actual value for this Output.
+        const applyDuringPreview = isKnown;
+
+        if (!applyDuringPreview) {
+            // We didn't actually run the function, our new Output is definitely
+            // **not** known.
+            return {
+                value: <U><any>undefined,
+                isKnown: false,
+                isSecret,
+            };
+        }
+    }
+
+    const transformed = await func(value);
+    if (Output.isInstance(transformed)) {
+        // Note: if the func returned a Output, we unwrap that to get the inner value
+        // returned by that Output.  Note that we are *not* capturing the Resources of
+        // this inner Output.  That's intentional.  As the Output returned is only
+        // supposed to be related this *this* Output object, those resources should
+        // already be in our transitively reachable resource graph.
+
+        // The callback func has produced an inner Output that may be 'known' or 'unknown'.
+        // We have to properly forward that along to our outer output.  That way the Outer
+        // output doesn't consider itself 'known' then the inner Output did not.
+        const innerValue = await transformed.promise();
+        const innerIsKnown = await transformed.isKnown;
+        const innerIsSecret = await (transformed.isSecret || Promise.resolve(false));
+
+        return {
+            value: innerValue,
+            isKnown: isKnown && innerIsKnown,
+            isSecret: isSecret || innerIsSecret,
+        };
+    } else {
+        // We successfully ran the inner function.  Our new Output should be considered known.
+        return {
+            value: transformed,
+            isKnown: true,
+            isSecret: false,
+        };
+    }
```

Comment:
note the lack of try/catch/finally.  It is totally fine to throw in this function for any reason.  If that happens, you will get an Output back at the end that is effectively 'rejected'.  All promise-vals on it will be in the rejected state, so later `awaits` in pulumi will cause them to be thrown the instant we ask about the `value`, `isKnown` or `isSecret` props.